src/compiler/checker.ts

```diff
@@ -439,7 +439,33 @@ module ts {
                     var declaration = forEach(result.declarations, d => isBlockOrCatchScoped(d) ? d : undefined);
 
                     Debug.assert(declaration !== undefined, "Block-scoped variable declaration is undefined");
-                    if (!isDefinedBefore(declaration, errorLocation)) {
+                    
+                    // first check if usage is lexically located after the declaration
+                    var isUsedBeforeDeclaration = !isDefinedBefore(declaration, errorLocation);
+                    if (!isUsedBeforeDeclaration) {
+                        // lexical check succedded however code still can be illegal.
+                        // - block scoped variables cannot be used in its initializers
+                        //   let x = x; // illegal but usage is lexically after definition
+                        // - in ForIn/ForOf statements variable cannot be contained in expression part
+                        //   for (let x in x)
+                        //   for (let x of x)
+
+                        // climb up to the variable declaration skipping binding patterns
+                        var variableDeclaration = <VariableDeclaration>getAncestor(declaration, SyntaxKind.VariableDeclaration);
+                        var container = getEnclosingBlockScopeContainer(variableDeclaration);
+
+                        if (variableDeclaration.parent.parent.kind === SyntaxKind.VariableStatement ||
+                            variableDeclaration.parent.parent.kind === SyntaxKind.ForStatement) {
+                            // variable statement/for statement case, use site should not be inside initializer
+                            isUsedBeforeDeclaration = isChildNode(errorLocation, variableDeclaration.initializer, container);
+                        }
+                        else if (variableDeclaration.parent.parent.kind === SyntaxKind.ForOfStatement || 
+                            variableDeclaration.parent.parent.kind === SyntaxKind.ForInStatement) {
+                                // ForIn/ForOf case - use site should not be used in expression part
+                            isUsedBeforeDeclaration = isChildNode(errorLocation, (<ForInStatement>variableDeclaration.parent.parent).expression, container);
+                        }

```

Comment:
Seems like neither of these checks handle initializers nested inside binding patterns:
```ts
let [x = x] = 0;
for (let [x = x] of []) { }
```